File: core/src/com/spacefist/managers/Manager.java
package com.spacefist.managers;

import com.badlogic.gdx.utils.Array;
import com.spacefist.GameData;
import com.spacefist.entities.Entity;
import org.jetbrains.annotations.NotNull;

import java.util.Iterator;

public class Manager<T extends Entity> implements Iterable<T> {

    protected Array<T> entities;
    protected GameData gameData;

    public Manager(GameData gameData) {
        this.gameData = gameData;
        entities      = new Array<T>(false, 16);
    }

    protected void add(@NotNull T entity) {
        assert entity != null;

        entities.add(entity);
    }

    protected void remove(@NotNull T entity) {
        assert entity != null;

        entities.removeValue(entity,true);
    }

    public void update() {
        for (T entity : entities) {
            if (entity.isAlive()) {
                entity.update();
            }
        }
    }

    public void draw() {
        for(T entity : entities) {
            entity.draw();
        }
    }

    @NotNull
    public Iterable<T> collisions(@NotNull Entity obj) {
        Array<T> collisions = new Array<T>(false, 16);

        for(T entity : entities) {
            if(entity.isAlive() && entity.getRectangle().overlaps(obj.getRectangle())) {
                collisions.add(entity);
            }
        }

        return collisions;
    }

    public void clear() {
        entities.clear();
    }

    @Override
    public Iterator<T> iterator() {
        return entities.iterator();
    }
}

File: core/src/com/spacefist/managers/Manager.kt
package com.spacefist.managers

import com.badlogic.gdx.utils.Array
import com.spacefist.GameData
import com.spacefist.entities.Entity

open class Manager<T : Entity>(protected var gameData: GameData) : Iterable<T> {

    protected var entities: Array<T> = Array(false, 16)

    protected fun add(entity: T) {
        entities.add(entity)
    }

    protected fun remove(entity: T) {
        entities.removeValue(entity, true)
    }

    open fun update() {
        entities
            .filter  { it.isAlive  }
            .forEach { it.update() }
    }

    fun draw() {
        entities.forEach { entity -> entity.draw() }
    }

    fun collisions(obj: Entity): Iterable<T> {
        val collisions = Array<T>(false, 16)

        for (entity in entities) {
            if (entity.isAlive && entity.rectangle!!.overlaps(obj.rectangle)) {
                collisions.add(entity)
            }
        }

        return collisions
    }

    fun clear() {
        entities.clear()
    }

    override fun iterator(): Iterator<T> {
        return entities.iterator()
    }
}
